src/org/parosproxy/paros/network/HttpSender.java

```diff
@@ -384,24 +386,35 @@ public void sendAndReceive(HttpMessage msg, boolean isFollowRedirect) throws IOE
 				sendAuthenticated(msg, isFollowRedirect);
 				return;
 			}
+			// POST/PUT method cannot be redirected by library. Need to follow by code
 
 			// ZAP: Reauthentication when sending a message from the perspective of a User
 			sendAuthenticated(msg, false);
 
+			// Update session and Cookie header if a Set-Cookie header has been received
+			notifyResponseListeners(msg);
+			HttpSession session = msg.getHttpSession();
+			if (session != null) {

```

Comment:
This should be done by `ExtensionHttpSessions`, this class shouldn't need to set/use the `HttpSession` directly.